components/org.wso2.carbon.identity.application.authenticator.hypr/src/main/java/org/wso2/carbon/identity/application/authenticator/hypr/HyprAuthenticator.java

```diff
@@ -0,0 +1,489 @@
+                    .addParameter("AuthenticatorName",
+                            HYPR.AUTHENTICATOR_FRIENDLY_NAME);
+
+            if (authenticationStatus != null) {
+                hyprLoginPageURLBuilder.addParameter("status", String.valueOf(authenticationStatus.getName()));
+                hyprLoginPageURLBuilder.addParameter(
+                        "message", String.valueOf(authenticationStatus.getMessage()));
+            }
+
+            String hyprLoginPageURL = hyprLoginPageURLBuilder.build().getAbsolutePublicURL();
+            response.sendRedirect(hyprLoginPageURL);
+
+        } catch (IOException e) {
+            throw new HYPRAuthnFailedException(
+                    ErrorMessages.AUTHENTICATION_FAILED_REDIRECTING_LOGIN_FAILURE.getCode(),
+                    ErrorMessages.AUTHENTICATION_FAILED_REDIRECTING_LOGIN_FAILURE
+                            .getMessage(), e);
+        } catch (URLBuilderException e) {
+            throw new HYPRAuthnFailedException(
+                    ErrorMessages.AUTHENTICATION_FAILED_BUILDING_LOGIN_URL_FAILURE.getCode(),
+                    ErrorMessages.AUTHENTICATION_FAILED_BUILDING_LOGIN_URL_FAILURE
+                            .getMessage(), e);
+        }
+    }
+
+    /**
+     * Send a push notification to the user-registered devices to start the user authentication process using
+     * HYPR, the external identity provider.
+     *
+     * @param request  The request that is received by the authenticator.
+     * @param response The response that is received to the authenticator.
+     * @param context  The Authentication context received by the authenticator.
+     * @throws AuthenticationFailedException Exception while sending push notification to the registered device.
+     */
+    private void initiateHYPRAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+                                                   AuthenticationContext context) throws AuthenticationFailedException {
+
+        try {
+
+            String username = request.getParameter(HYPR.USERNAME);
+            String sessionDataKey = request.getParameter(HYPR.SESSION_DATA_KEY);
+
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Initiated the authentication process for the user : " + username);
+            }
+
+            HYPRAuthorizationAPIClient hyprAuthorizationAPIClient = getHYPRAPIClient(context);
+
+            // Get the registered devices.
+            ArrayNode registeredDevices = getRegisteredDevices(username, hyprAuthorizationAPIClient);
+
+            // If an empty array received for the registered devices redirect user back to the login page and
+            // display "Invalid username" since a HYPR user cannot exist without a set of registered devices.
+            if (registeredDevices.isEmpty()) {
+                redirectHYPRLoginPage(response, sessionDataKey, HYPR.AuthenticationStatus.INVALID_REQUEST);
```

Comment:
As the registered device means, whether the user is entered an incorrect username or the user doesn't exist. In this case what is the error message that we prompt for user?

Reply:
The prompted message : "Invalid username provided"